kousa/lib/kousa/utils/hash.ex

```diff
@@ -1,21 +1,29 @@
 defmodule Kousa.Utils.Hash do
   alias Argon2
+  alias IO
 
-  @type hash_result :: String.t() | :error
+  @type hash_result :: String.t() | {:error, String.t()}
 
   @type hash_check :: {:ok, hash_result} | {:error, String.t()}
 
   @spec hash_ip(String.t()) :: hash_result
   def hash_ip(ip) do
-    Argon2.hash_pwd_salt(ip)
+    if is_nil(ip) do
+      {:error, "ip is nil"}
+    else
+      # Should we refactor this so we don't pipe a function call?
+      # Encodes Hash into Base16
+      :crypto.mac(:hmac, :sha256, Application.fetch_env!(:kousa, :ip_hashing_key_base), ip)
```

Comment:
This isn't a good implementation.  It would be very easy to get the base key and create a rainbow table for the whole IP space. This methodology uses ip_hashing_key_base as "pepper".  To do this method properly, it should be done in conjunction with salts.  Looking at [RFC7518](https://datatracker.ietf.org/doc/html/rfc7518#section-3.2), the uniquely created salt should two time the length of the input (this would  mean 2*45 = 90 characters). The hash can be stored in the same field as the IP using the [crypt)](https://en.wikipedia.org/wiki/Crypt_(C)) format.

Reply:
I thought the whole point is someone wouldn't have the hash key because its a secret

Reply:
The key should not be a substitute for a good salt. I think that this approach shares the same issues as a shared [pepper](https://en.wikipedia.org/wiki/Pepper_(cryptography)#Shared_Secret_Pepper).
Looking at it from the prospective of a hacker or an evil company, this value is store via environment variables `System.get_env("IP_HASHING_KEY_BASE")`. Being that value is constant (shared for all hashing) and can be fetched by both a company or hacker if they gain access to the system. It won't take much more work for either entity to create a rainbow table using this know secret (presuming they managed to get database access). Every IP is bijective with the entries of a rainbow table for a know key. The number of IP that exists in the database, denoted as **_n_**, is only a subset of the know rainbow table entity. So let's take all the Class C IPs, denoted as **_c_**, to create a valid rainbow table, the IP space would need to be computed for each salt. This will take the complexity from O(**_c_**) to O(**_c_*****_n_**). Don't get me wrong, I think having anything is better than no hashing. I just think the approach could be broken based on  the methods displayed.

Reply:
how about generating and encrypting a secret private key using a encryption key and storing that private key in the database alongside the user? this method forces an attacker to get access to the database **and** the environment variables.